### ISSUE TITLE
Extra Output Variables in v8.7

### ISSUE THREAD

**Issue by jonwinkler** [NREL/EnergyPlus#6106]
Despite including a Key Value in certain Output:Variable objects, all corresponding output variables are being written (as if an * was included for the key value). This does not occur in v8.6.

This only occurs for certain outputs. See attached IDF and CSV files. For example, the Zone Mean Air Temperature is being written for all zones when it should only be writing it for 'living_1'. This is not happening for all output variables but I've discovered two so far (see output in attached ZIP file).

Attached is ZIP file containing two IDFs and two CSV files. A diff on the IDFs show they're the same file excluding the IDD changes from 8.6 to 8.7. The output from 8.7 has extra unwanted columns. This an issue when running 10,000's of simulations when output file size is a concern.

@nmerket 

[files.zip](https://github.com/NREL/EnergyPlus/files/944660/files.zip)

**Comment by mjwitte** [comment]
@mbadams5

**Comment by mbadams5** [comment]
At first glance, it seems like it is adding a * key value for certain output variables and the culprit is [here](https://github.com/NREL/EnergyPlus/blob/78a111df4adc0ac10b929e2abd28ac9391d46bfa/src/EnergyPlus/InputProcessor.cc#L5480). These are added due to the following in your input file which then causes all those keys to be added internally.
```
Output:Table:SummaryReports,
   AllSummaryAndMonthly;
```

What I haven't figured out yet, is why E+ 8.6 did not have this issue. The new code in E+ 8.7 is working properly, it just "thinks" there are extra Output:Variables added to the IDF. I'll keep looking into this more and try to find a solution.

**Comment by mjwitte** [comment]
@JasonGlazer ?

**Comment by mjwitte** [comment]
@mbadams5 I can't help thinking the problem is that the re2 matching is being applied at the wrong place and it's expanding in the wrong direction when it sees the "*" in `OutputVariablesForSimulation` rather than expanding based on `ReqRepVars`.

**Comment by mbadams5** [comment]
@mjwitte I don't fully understand what you mean when you say `it's expanding in the wrong direction when it sees the "*" in OutputVariablesForSimulation rather than expanding based on ReqRepVars.`
The regex matching is being done in the same exact place as the previous linked list search and as far as I am aware is doing the exact same matching, but obviously something is different.

**Comment by mjwitte** [comment]
@mbadams5 Neither do I - just fishing for an explanation.  I don't fully understand all the output setup functions, but what I see at this point are essentially three passes (figuring this out as I go):

1. To activate accumulation of all the required values (and avoid accumulating variables that are never used for anything).  This pass goes through idf output variable requests in [InputProcessor::PreScanReportingVariables](https://github.com/NREL/EnergyPlus/blob/78a111df4adc0ac10b929e2abd28ac9391d46bfa/src/EnergyPlus/InputProcessor.cc#L5257) and then it calls [InputProcessor::AddVariablesForMonthlyReport](https://github.com/NREL/EnergyPlus/blob/78a111df4adc0ac10b929e2abd28ac9391d46bfa/src/EnergyPlus/InputProcessor.cc#L5437) which you identified earlier.  Both of these  call [InputProcessor::AddRecordToOutputVariableStructure](https://github.com/NREL/EnergyPlus/blob/78a111df4adc0ac10b929e2abd28ac9391d46bfa/src/EnergyPlus/InputProcessor.cc#L5920) which ultimately sets up `DataOutputs::OutputReportingVariables`.  At this point, the KeyValues in `OutputReportingVariables` include *s or strings from idf output objects.

2. The next pass sets up `OutputProcessor::ReqRepVars` with a list of requested outputs from the idf objects.  This happens in [OutputProcessor::GetReportVariableInput](https://github.com/NREL/EnergyPlus/blob/78a111df4adc0ac10b929e2abd28ac9391d46bfa/src/EnergyPlus/OutputProcessor.cc#L806).  Important to note here that if the idf request has "*" for the key, `ReqRepVars( Loop ).Key` is set to a blank string, so later this is checked to see if it's empty or not.

3. The final pass happens whenever [OutputProcessor::SetUpOutputVariable](https://github.com/NREL/EnergyPlus/blob/78a111df4adc0ac10b929e2abd28ac9391d46bfa/src/EnergyPlus/OutputProcessor.cc#L5060).  `SetUpOutputVariable` is called once for every possible unique key for a given variable name.  So when it calls [CheckReportVariable](https://github.com/NREL/EnergyPlus/blob/78a111df4adc0ac10b929e2abd28ac9391d46bfa/src/EnergyPlus/OutputProcessor.cc#L552) it is passing a "hard" string in `KeyedValue` that expects to find either an exact match or a "*" in OutputReportingVariables.  If either of these conditions are true, then that variable is added to ` with some extra work to allow to multiple reporting frequencies and avoid duplicates.

    a. First, it checks to see if the variable name is in the list, if not, we're done.
    b. Then it calls `BuildKeyVarList` and `AddBlankKeys`.
    c.  [BuildKeyVarList](https://github.com/NREL/EnergyPlus/blob/78a111df4adc0ac10b929e2abd28ac9391d46bfa/src/EnergyPlus/OutputProcessor.cc#L653)  loops through `ReqRepVars` skipping any that have a blank key (those are handled in `AddBlankKeys`).  Then it calss [DataOutputs::FindItemInVariableList](https://github.com/NREL/EnergyPlus/blob/cecaea1295fd603c9ae24d89a63339bd45fb454e/src/EnergyPlus/DataOutputs.cc#L147) to check for a match.  At this point `KeyedValue` is a full name passed in from the call to `SetUpOutputVariable`.  

I think I see the problem.  In `BuildKeyVarList` prior to the regex changes, [this line](https://github.com/NREL/EnergyPlus/blob/ff1514acdb773b8a6290b6b4bfed936f86d6b2d4/src/EnergyPlus/OutputProcessor.cc#L700) is comparing `KeyedValue` with `ReqRepVars`.

```
if ( ! SameString( ReqRepVars( Loop ).Key, KeyedValue ) ) continue;
```

But the [same place in the current code](https://github.com/NREL/EnergyPlus/blob/cecaea1295fd603c9ae24d89a63339bd45fb454e/src/EnergyPlus/OutputProcessor.cc#L701) is calling `FindItemInVariableList` which is comparing against `OutputReportingVariables` instead of `ReqRepVars`.

```
if ( ! DataOutputs::FindItemInVariableList( KeyedValue, VariableName ) ) continue;
```

So, I think the fix is to write a similar function to `FindItemInVariableList` that searches  `ReqRepVars`, or change `FindItemInVariableList` to have another argument that passes a pointer to the desired search array.

**Comment by Myoldmopar** [comment]
Closed via #6303
